pkg/controller/smart-clone-controller.go

```diff
@@ -0,0 +1,329 @@
+		} else {
+			runtime.HandleError(errors.Errorf("Error getting Snapshot: %s", name))
+			return false
+		}
+	}
+
+	if pvc != nil && pvc.Status.Phase == corev1.ClaimBound {
+		if pvc.ObjectMeta.Annotations[AnnSmartCloneRequest] == "true" {
+			snapshotName := pvc.Spec.DataSource.Name
+			snapshotToDelete, err := c.snapshotsLister.VolumeSnapshots(ns).Get(snapshotName)
+			if err != nil {
+				runtime.HandleError(errors.Errorf("Error getting Snapshot: %s", snapshotName))
+				return false
+			}
+			if snapshotToDelete != nil {
+				klog.V(3).Infof("ProcessNextItem snapshotName: %s", snapshotName)
+				err = c.csiClientSet.SnapshotV1alpha1().VolumeSnapshots(ns).Delete(snapshotName, &metav1.DeleteOptions{})
+				if err != nil {
+					klog.Errorf("error deleting snapshot for smart-clone %q: %v", key, err)
+					return true
+				}
+				klog.V(3).Infof("Snapshot deleted: %s", snapshotName)
+				anno := map[string]string{}
+				anno[AnnPodPhase] = string(corev1.PodSucceeded)
+				lab := map[string]string{}
+				if !checkIfLabelExists(pvc, common.CDILabelKey, common.CDILabelValue) {
+					lab = map[string]string{common.CDILabelKey: common.CDILabelValue}
+				}
+				pvc, err = updatePVC(c.clientset, pvc, anno, lab)
+				if err != nil {
+					klog.Errorf("could not update pvc %q annotation and/or label", err)
+					return false
+				}
+			}
+		}
+	} else if snapshot != nil {
+		err := c.syncSnapshot(key.(string))
+		if err != nil {
+			klog.Errorf("error processing snapshot %q: %v", key, err)
+			return true
+		}
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextSnapshotItem: processing pvc/snapshot %q completed", key))
+}
+
+func (c *SmartCloneController) syncSnapshot(key string) error {
+	snapshot, exists, err := c.snapshotFromKey(key)
+	if err != nil {
+		return err
+	} else if !exists {
+		c.podExpectations.DeleteExpectations(key)
+	}
+
+	if snapshot == nil {
+		return nil
+	}
+	_, ok := snapshot.Annotations[AnnSmartCloneRequest]
+	if !ok {
+		//ignoring snapshot, not created by DataVolume Controller
+		return nil
+	}
+
+	snapshotReadyToUse := snapshot.Status.ReadyToUse
+	klog.V(3).Infof("Snapshot \"%s/%s\" - ReadyToUse: %t", snapshot.Namespace, snapshot.Name, snapshotReadyToUse)
+	if !snapshotReadyToUse {
+		//set Datavolume phase to Scheduled
+		return nil
+	}
+
+	return c.processNextSnapshotItem(snapshot)
+}
+
+// Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
+// processing this pvc, which would result in multiple pods for the same pvc.
+func (c *SmartCloneController) processNextSnapshotItem(snapshot *csisnapshotv1.VolumeSnapshot) error {
+	// anno := map[string]string{}
+
+	snapshotKey, err := cache.MetaNamespaceKeyFunc(snapshot)
+	if err != nil {
+		return err
+	}
+
+	// expectations prevent us from creating multiple pods. An expectation forces
+	// us to observe a pod's creation in the cache.
+	needsSync := c.podExpectations.SatisfiedExpectations(snapshotKey)
```

Comment:
Oh I see, we need them, they are just named podExpectations, instead of snapshotExpectations.

Reply:
Done, renamed.